File: source/FuerzaBruta.cpp
#include "FuerzaBruta.h"
#include <limits>
#include <vector>


// función recursiva
void encontrarSeamFuerzaBrutaRec(const std::vector<std::vector<double>>& energia, int i,int j,int n ,int m ,std::vector<int>&S ,std::vector<int>&B, double energiaActual, double& mejorEnergia) {
    if(i==n){               // caso base: es el final de la imagen 
        if(B.empty() || energiaActual < mejorEnergia ){         // si encontró un camino mejor actualiza y guarda ese camino
            mejorEnergia = energiaActual;
            B = S;
        }
        return;
    }

    if(j>0){                // baja por la rama izquierda siempre y cuando no esté en un borde
        S.push_back(j - 1);
        encontrarSeamFuerzaBrutaRec(energia, i + 1, j - 1, n, m, S, B, energiaActual + energia[i][j - 1], mejorEnergia);
        S.pop_back();
    }
    
    S.push_back(j);         // en todos los casos baja por la rama del medio
    encontrarSeamFuerzaBrutaRec(energia, i + 1, j, n, m, S, B, energiaActual + energia[i][j], mejorEnergia);
    S.pop_back();

    if(j<m-1){             // baja por la rama derecha siempre y cuando no esté en un borde
        S.push_back(j + 1);
        encontrarSeamFuerzaBrutaRec(energia, i + 1, j + 1, n, m, S, B, energiaActual + energia[i][j + 1], mejorEnergia);
        S.pop_back();
    }
}


// función principal
std::vector<int> encontrarSeamFuerzaBruta(const std::vector<std::vector<double>>& energia) {
    if (energia.empty()) {
        return {};
    }

    int n= energia.size();          // filas
    int m= energia[0].size();       // columnas

    std::vector<int> B;             // mejor camino encontrado
    double mejorEnergia = std::numeric_limits<double>::infinity();

    for (int j = 0; j < m; j++) {   // comenzando desde cada columna posible en la primer fila
        std::vector<int> S;
        S.push_back(j);
        encontrarSeamFuerzaBrutaRec(energia, 1, j, n, m, S, B, energia[0][j], mejorEnergia);
    }

    for (int i = 0; i < B.size(); i++) {
        B[i] += 1;                  // pasa de base 0 a base 1 (formato de salida)
    }
    return B;
}

File: codigoPython/FuerzaBruta.py
#Cambios necesarios:
#El único cambio estructural necesario: mejorEnergia en C++ es double& (referencia modificable). 
#En Python los floats son inmutables, así que lo envolví en una lista de un elemento (mejorEnergia[0]) para replicar exactamente ese comportamiento.
#La lógica es idéntica.



# función recursiva
def encontrarSeamFuerzaBrutaRec(energia: list[list[float]], i: int, j: int, n: int, m: int, S: list[int], B: list[int], energiaActual: float, mejorEnergia: list[float]) -> None:
    if i == n:              # caso base: es el final de la imagen
        if not B or energiaActual < mejorEnergia[0]:        # si encontró un camino mejor actualiza y guarda ese camino
            mejorEnergia[0] = energiaActual
            B[:] = S[:]
        return

    if j > 0:               # baja por la rama izquierda siempre y cuando no esté en un borde
        S.append(j - 1)
        encontrarSeamFuerzaBrutaRec(energia, i + 1, j - 1, n, m, S, B, energiaActual + energia[i][j - 1], mejorEnergia)
        S.pop()

    S.append(j)             # en todos los casos baja por la rama del medio
    encontrarSeamFuerzaBrutaRec(energia, i + 1, j, n, m, S, B, energiaActual + energia[i][j], mejorEnergia)
    S.pop()

    if j < m - 1:           # baja por la rama derecha siempre y cuando no esté en un borde
        S.append(j + 1)
        encontrarSeamFuerzaBrutaRec(energia, i + 1, j + 1, n, m, S, B, energiaActual + energia[i][j + 1], mejorEnergia)
        S.pop()

  # función principal
def encontrarSeamFuerzaBruta(energia: list[list[float]]) -> list[int]:
    if not energia:
        return []

    n = len(energia)        # filas
    m = len(energia[0])     # columnas

    B = []                          # mejor camino encontrado
    mejorEnergia = [float('inf')]

    for j in range(m):      # comenzando desde cada columna posible en la primer fila
        S = []
        S.append(j)
        encontrarSeamFuerzaBrutaRec(energia, 1, j, n, m, S, B, energia[0][j], mejorEnergia)
    
    return B
